qiskit/quantum_info/mitigation/base_readout_mitigator.py

```diff
@@ -0,0 +1,156 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Base class for readout error mitigation.
+"""
+
+from abc import ABC, abstractmethod
+from typing import Optional, List, Dict, Iterable, Tuple, Union
+import numpy as np
+from qiskit.result import Counts
+
+try:
+    import matplotlib.pyplot as plt
+
+    _HAS_MATPLOTLIB = True
+except ImportError:
+    _HAS_MATPLOTLIB = False
+
+
+class BaseReadoutMitigator(ABC):
+    """Base readout error mitigator class."""
+
+    def quasi_probabilities(self,
+                            data: Counts,
+                            qubits: Iterable[int] = None,
+                            shots: Optional[int] = None
+    ) -> (Dict[str, float], Dict[str, float]):
+        """Convert counts to a dictionary of quasi-probabilities
+
+        Args:
+            data: Counts to be mitigated.
+            qubits: the physical qubits measured to obtain the counts clbits.
+                    If None these are assumed to be qubits [0, ..., N-1]
+                    for N-bit counts.
+            shots: Optional, the total number of shots, if None shots will
+                be calculated as the sum of all counts.
+
+        Returns:
+            A dictionary containing pairs of [output, mean] where "output"
+                is the key in the dictionaries,
+                which is the length-N bitstring of a measured standard basis state,
+                and "mean" is the mean of non-zero quasi-probability estimates.
+            A dictionary containing pairs of [output, standard deviation]
+                where "output" is the key in the dictionaries,
+                which is the length-N bitstring of a measured standard basis state,
+                and "standard deviation" is the standard deviation of the non-zero
+                quasi-probability estimates.
+
+        Raises:
+            QiskitError: if qubits is not None and does not match the number of count clbits.
+        """
+
+    def expectation_value(self,
+                          data: Counts,
+                          diagonal: np.ndarray,
```

Comment:
This cannot be an array type or it does not scale.  Either a `dict` or `callable` should be here.  I think the dict is probably the better choice, but it is debatable.